### ISSUE TITLE
New table idea: bond configuration

### ISSUE THREAD

**Issue by skottler** [facebook/osquery#1173]
We extensively use bonding and it'd be extremely useful to be able to get the members of a bond as well as the state of interfaces in a bond. Since bond configuration lives in the interface configuration, but the status of each bond is in `/sys` it could be useful to provide a new table for this kind of information since the format for bond status is not like much of the rest of `/sys. There are two options here:

* Add a new table with columns specifically tied to bond configuration.
* Append bond information to interfaces.

There doesn't really seem to be a right answer here and I don't have a strong preference. Is there anything along the lines of existing patterns for similar problems that might be useful here for figuring out the way forward? I'm happy to implement this feature.

**Comment by theopolis** [comment]
Hey @skottler, good questions. We don't mind duplicating data in tables, normalization doesn't really mean anything in this case since we're mostly force fitting concepts into tabular form. But optimizing for obvious foreign keys (like pid or interface name) is supported and encouraged. Then we can separate concepts like process metadata and various tabular-esqe process supplemental information like environment variables, open sockets, file descriptors, memory maps, etc. Using that [little bit of optimization](https://github.com/facebook/osquery/blob/master/osquery/tables/system/linux/processes.cpp#L265), `JOIN`s make sense!

To represent bonds we could definitely use interface names as a foreign key.

I'm **pretty** sure there's a 1-1 relationship between bond and interface? I can't put an interface into multiple bonds, right?

I'd vote for an `interface_bonds` table and the primary key is the bond name and columns are bond stats/metadata. Then we add a `bond` column to `interface_details`, as a foreign key to `interface_bonds`.

**Comment by skottler** [comment]
> I'm pretty sure there's a 1-1 relationship between bond and interface? I can't put an interface into multiple bonds, right?

In the case of the `bond-mode` being `active-passive` there is a 1-1 relationship between the `bond-master` and interface, there's only a single master with potentially several slaves. For example (with some address redacted):

```
auto lo
iface lo inet loopback

auto eth0
iface eth0 inet manual
        bond-master bond0
        bond-primary eth0

auto eth1
iface eth1 inet manual
        bond-master bond0

auto bond0
iface bond0 inet manual
        bond-mode active-backup
        bond-miimon 100
        bond-slaves none
        address 10.32.1.1
        netmask 255.255.0.0
        gateway 10.32.0.1
```

`/proc/net/bonding/bond0` then has the following state about the bond:

```
Bonding Mode: fault-tolerance (active-backup)
Primary Slave: eth0 (primary_reselect always)
Currently Active Slave: eth0
MII Status: up
MII Polling Interval (ms): 100
Up Delay (ms): 0
Down Delay (ms): 0

Slave Interface: eth0
MII Status: up
Speed: 40000 Mbps
Duplex: full
Link Failure Count: 0
Permanent HW addr: 00:00:00:00:00:00
Slave queue ID: 0

Slave Interface: eth1
MII Status: up
Speed: 40000 Mbps
Duplex: full
Link Failure Count: 1
Permanent HW addr: 00:00:00:00:00:00
Slave queue ID: 0
```

Of course the MAC and other details on the interfaces which are part of the bond are already in `interface_details`, but things like `Currently Active Slave`, `MII Status`, or `Link Failure Count` would be useful to exist in the potentially new `interface_bonds` table. This seems to be a reasonable way forward, I'm gonna start working on a pull request for this.

Thanks and have a great weekend!

**Comment by theopolis** [comment]
Perfect! I could see expressing that info in an `interface_bonds` table like:

| bond | interface | mode | mii_status | speed | duplex | link_failures | queue_id | primary |
|---------|-------------|---------|---------------|----------|-----------|-----------------|--------------|------------|
| bond0 | eth0 | active-backup | up | 40000 | full | 0 | 0 | 1 |
| bond0 | eth1 | active-backup | up | 40000 | full | 1 | 1 | 0 |

(The `primary` boolean column goes off the right side of the github-comment, you'll need to horizontal scroll.)

There a bit of repeated information, but that's no big deal. You can see with a simple schedule of the bond interfaces table (`SELECT * FROM interface_bonds`) whenever interfaces leave/join any bond, and when status/counts change.

Thanks @skottler, you do the same, and definitely looking forward to the PR! :dancers:

**Comment by marpaia** [comment]
I added this to the list in #619 with a reference to this issue. I'm closing this, so that it doesn't show up in the open issues list, but conversation on the topic should still occur here.

**Comment by xavinux** [comment]
@marpaia @theopolis think in the spec file interface_bonds.table like this:

table_name("interface_bonds")
description("Displays information about network interfaces used for bonding.")
schema([
    Column("bond", TEXT, "Name of the bond"),
    Column("interface", TEXT, "Name of the network interfaces used for the bonding"),
    Column("mode", TEXT, "Indicates the Bonding mode"),
    Column("mii_status", TEXT, "Check the status of the network interfaces"),
    Column("speed", BIGINT, "Check the link speed of the network interfaces"),
    Column("duplex", TEXT, "Check if network interfaces are capable of carrying signals in both directions"),
    Column("link_failures", INTEGER, "0 (false) and 1 (true)"),
    Column("queue_id", INTEGER, "0 (false) and 1 (true)"),
    Column("primary", INTEGER, "0 (false) and 1 (true)"),
])
implementation("linux/interface_bonds@genInterfaceBonds")

Located at /osquery/specs/linux/interface_bonds.table

**Comment by xavinux** [comment]
@theopolis @marpaia have created this table named interface_bonds like this:

Column "master_bonds" --> Name of the master bonds.
Column "interfaces"  --> Name of the network interfaces used for the bonding.
Column "mode"  --> Indicates the Bonding mode.
Column "mii_status" --> Check the status of the network interfaces.
Column "bonds_speed"  --> Check the link speed of bonding.
Column "duplex"  --> Check if network interfaces are capable of carrying signals in both directions.
Column "tx_errors"  --> Errors in bonds transmition.
Column "rx_errors"  --> Errors in bonds reception.
Column "tx_bytes"  --> Bytes transmitted by the bonds.
Column "rx_bytes"  --> Bytes received by the bonds.
Column "queue_id"  --> The queue id for the slaves.
Column "primary"  --> Name of the primary interface.
Column "active_slave"  --> Name of the active slave interface.

Has not been finished yet as I'm working on the .cpp file, but compilation runs ok on Ubuntu Server 14.04 64 bits and this new table is displayed:

./osquery/build/linux/osquery/osqueryi

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
osquery - being built, with love, at Facebook

Using a virtual database. Need help, type '.help'
osquery> .tables
 ....
  => interface_addresses
  => interface_bonds
  => interface_details
....

osquery> select * from interface_bonds;
+--------------+------------+-----------------+------------+-------------+--------+-----------+-----------+----------+----------+---------------+---------+--------------+
| master_bonds | interfaces | mode            | mii_status | bonds_speed | duplex | tx_errors | rx_errors | tx_bytes | rx_bytes | queue_id      | primary | active_slave |
+--------------+------------+-----------------+------------+-------------+--------+-----------+-----------+----------+----------+---------------+---------+--------------+
| bond0        | eth0 eth1  | active-backup 1 | up         | 1000        | full   | 0         | 0         | 602086   | 1444143  | eth0:0 eth1:0 | eth0    | eth0         |
+--------------+------------+-----------------+------------+-------------+--------+-----------+-----------+----------+----------+---------------+---------+--------------+

**Comment by theopolis** [comment]
Looking good! I'd recommend renaming the `master_bonds` column to `name` or `bond`. The same with `bonds_speed`, shorten this to `speed`. The primary interface is good to know, maybe reorder the column list to show that next to the bond name?

**Comment by theopolis** [comment]
Do you have a set of commits you could open a pull request with so we can review the code?

**Comment by xavinux** [comment]
I've reorder and rename some columns like this:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./build/linux/osquery/osqueryi

osquery - being built, with love, at Facebook

Using a virtual database. Need help, type '.help'
osquery> select * from interface_bonds;
+-------+---------+------------+-----------------+------------+-------+--------+-----------+-----------+----------+----------+---------------+--------------+
| bond  | primary | interfaces | mode            | mii_status | speed | duplex | tx_errors | rx_errors | tx_bytes | rx_bytes | queue_id      | active_slave |
+-------+---------+------------+-----------------+------------+-------+--------+-----------+-----------+----------+----------+---------------+--------------+
| bond0 | eth0    | eth0 eth1  | active-backup 1 | up         | 1000  | full   | 0         | 0         | 602086   | 1444143  | eth0:0 eth1:0 | eth0         |
+-------+---------+------------+-----------------+------------+-------+--------+-----------+-----------+----------+----------+---------------+--------------+
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

I've finished the interface_bonds.table spec file and would like you to review my code.
If you think that there is something else that can be improve, please let it me know.

**Comment by xavinux** [comment]
@theopolis have added two new columns:
"ttype" next to the "mode" column, which displays the bonding number.
"miimon" which displays the frequency in milliseconds of MII (Media Independent Interface) link monitoring:


 ./build/linux/osquery/osqueryi
osquery - being built, with love, at Facebook
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Using a virtual database. Need help, type '.help'
osquery> 
osquery> select * from interface_bonds;
+-------+---------+------------+---------------+------+------------+-------+--------+-----------+-----------+----------+----------+----------------+--------------+--------+
| bond  | primary | interfaces | mode          | type | mii_status | speed | duplex | tx_errors | rx_errors | tx_bytes | rx_bytes | queue_id       | active_slave | miimon |
+-------+---------+------------+---------------+------+------------+-------+--------+-----------+-----------+----------+----------+----------------+--------------+--------+
| bond0 | eth0    | eth0 eth1
 | active-backup | 1    | up         | 1000  | full   | 0         | 0         | 7558099  | 13842015 | eth0:0 eth1:0
 | eth0         | 100    |
+-------+---------+------------+---------------+------+------------+-------+--------+-----------+-----------+----------+----------+----------------+--------------+--------+
